[CI] SearchDirectoryTests testGetHighestOffsetSegmentInfos failing

**Build Scans:**
- [elasticsearch-serverless-intake #7107 / java_run_serverless_elasticsearch_functional_tests](https://gradle-enterprise.elastic.co/s/pdmg44f26wozw)
- [elasticsearch-serverless-es-pr-check #26484 / java_run_serverless_elasticsearch_functional_tests](https://gradle-enterprise.elastic.co/s/midaj6xmuqgwe)
- [elasticsearch-serverless-es-pr-check #26437 / java_run_serverless_elasticsearch_functional_tests](https://gradle-enterprise.elastic.co/s/misayimhdpvbu)
- [elasticsearch-serverless-es-pr-check #26407 / java_run_serverless_elasticsearch_functional_tests](https://gradle-enterprise.elastic.co/s/yotxuz6jaxdfc)
- [elasticsearch-serverless-es-pr-check #26179 / java_run_serverless_elasticsearch_functional_tests](https://gradle-enterprise.elastic.co/s/iszftqoclfo5s)
- [elasticsearch-serverless-es-pr-check #26004 / java_run_serverless_elasticsearch_functional_tests](https://gradle-enterprise.elastic.co/s/yf3swkgpzlax6)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:stateless:test" --tests "org.elasticsearch.xpack.stateless.lucene.SearchDirectoryTests.testGetHighestOffsetSegmentInfos" -Dtests.seed=CC4E9FEAD977CA9C -Dtests.locale=gu -Dtests.timezone=Africa/Ndjamena -Druntime.java=24
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(co.elastic.elasticsearch.stateless.lucene.SearchDirectoryTests),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!(testGetHighestOffsetSegmentInfos),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: expected:<51> but was:<268>
```

**Issue Reasons:**
- [main] 6 failures in test testGetHighestOffsetSegmentInfos (1.8% fail rate in 335 executions)
- [main] 6 failures in step java_run_serverless_elasticsearch_functional_tests (2.3% fail rate in 257 executions)
- [main] 5 failures in pipeline elasticsearch-serverless-es-pr-check (2.1% fail rate in 233 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

---

Migrated from https://github.com/elastic/elasticsearch-serverless/issues/3923

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -344,9 +344,6 @@ tests:
 - class: org.elasticsearch.xpack.stateless.engine.translog.StatelessTranslogIT
   method: testTranslogMasterFailureOnlyStressRecoveryTest
   issue: https://github.com/elastic/elasticsearch/issues/147835
-- class: org.elasticsearch.xpack.stateless.lucene.SearchDirectoryTests
-  method: testGetHighestOffsetSegmentInfos
-  issue: https://github.com/elastic/elasticsearch/issues/147837
 - class: org.elasticsearch.xpack.stateless.cache.StatelessOnlinePrewarmingIT
   method: testShardPrewarming
   issue: https://github.com/elastic/elasticsearch/issues/147838

--- a/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/lucene/SearchDirectoryTests.java
+++ b/x-pack/plugin/stateless/src/test/java/org/elasticsearch/xpack/stateless/lucene/SearchDirectoryTests.java
@@ -383,10 +383,10 @@ public class SearchDirectoryTests extends ESTestCase {
                         searchDirectory.shardId,
                         blobLocations,
                         List.of(
-                            randomAlphaOfLength(10),
-                            randomAlphaOfLength(10),
-                            randomAlphaOfLength(10),
-                            randomAlphaOfLength(10),
+                            randomNonSiFileName(),
+                            randomNonSiFileName(),
+                            randomNonSiFileName(),
+                            randomNonSiFileName(),
                             // last file is the segment info so the last location should be returned
                             randomAlphaOfLength(10) + LuceneFilesExtensions.SI.getExtension()
                         )
@@ -414,10 +414,10 @@ public class SearchDirectoryTests extends ESTestCase {
                             blobLocations,
                             List.of(
                                 randomAlphaOfLength(10) + LuceneFilesExtensions.SI.getExtension(),
-                                randomAlphaOfLength(10),
-                                randomAlphaOfLength(10),
-                                randomAlphaOfLength(10),
-                                randomAlphaOfLength(10)
+                                randomNonSiFileName(),
+                                randomNonSiFileName(),
+                                randomNonSiFileName(),
+                                randomNonSiFileName()
                             )
                         )
                     );
@@ -429,6 +429,11 @@ public class SearchDirectoryTests extends ESTestCase {
                 }
             }
         }
+    }
+
+    private String randomNonSiFileName() {
+        String siExtension = LuceneFilesExtensions.SI.getExtension();
+        return randomValueOtherThanMany(name -> name.endsWith(siExtension), () -> randomAlphaOfLength(10));
     }
 
     public void testUpdateCommitWithReplicatedContent() throws IOException {

